### ISSUE TITLE
Memory leak using AMP and transformers

### ISSUE THREAD

**Issue by edenlightning** [PyTorchLightning/pytorch-lightning#4614]
Context can be found here: https://github.com/huggingface/transformers/issues/8403

10x memory consumption increase with native amp on PL 

I'm running under debugger pt15+apex side by side with pt16+native amp, https://github.com/huggingface/transformers/issues/8403
and found the first issue of objects not being freed

I ruled out GradScaler I think there is a huge memory leak somewhere.
Typically this happens due to circular references, but I don't know PL codebase... so asking for help.

Oh, and I ruled out transformers  - the HF trainer using the same model code doesn't have this issue.

**Comment by stas00** [comment]
So the first place where it fails to clear cuda cache when running with native amp is here: https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/connectors/checkpoint_connector.py#L62 (with apex it clears about 1/3rd of preallocated gpu ram at this point). The trace is:

```
restore_weights, checkpoint_connector.py:64
setup_training, training_loop.py:174
train, gpu_accelerator.py:51
fit, trainer.py:444
generic_train, lightning_base.py:398
main, finetune.py:413
<module>, finetune.py:446
```

**Comment by stas00** [comment]
OK, this one I sorted out not to be the culprit as I initially thought it was. 

The problem with native amp is that it doesn't cast to fp16 until much later in the game, hence the discrepancy. If for the sake of debug I force `.half()` here:
```
    def restore_weights(self, model: LightningModule):
[...]
        model.half()
        import gc
        gc.collect()
        # clear cache before restore
        if self.trainer.on_gpu:
            torch.cuda.empty_cache()
```
the cache gets emptied - identically to when apex amp is used - the model sizes on gpu are identical at this point. 

So this wasn't it. back to the drawing board.

**Comment by stas00** [comment]
I don't think autocast is doing anything, despite the calls being made through it's context manager. 

If I look at the forward inputs with apex, they are of dtype `float16` - with native they are `float32`

apex (see lower left corner)
![fp-16-apex-inputs](https://user-images.githubusercontent.com/10676103/98770962-366c4b00-2398-11eb-9289-388b4c1b25d9.png)

native (see lower left corner)
![fp-16-native-inputs](https://user-images.githubusercontent.com/10676103/98770976-4126e000-2398-11eb-9d15-0e07807554ae.png)

**Comment by stas00** [comment]
Reading more on autocast - it doesn't just autocast any operation, but only a select list of pytorch ops

https://pytorch.org/docs/stable/amp.html#ops-that-can-autocast-to-float16

> __matmul__, addbmm, addmm, addmv, addr, baddbmm, bmm, chain_matmul, conv1d, conv2d, conv3d, conv_transpose1d, conv_transpose2d, conv_transpose3d, GRUCell, linear, LSTMCell, matmul, mm, mv, prelu, RNNCell

So I was checking initially in the wrong parts of the `forward()` - where none of these have been used. So I rechecked again with `forward()` that contained some `linear` layers - same thing, the weights of `linear` remain `float32` and the result from `nn.Linear.forward()` is still `float32`.

**Comment by stas00** [comment]
OK, so I did find a few lines of code where it did `autocast` to `float16` [here](https://github.com/huggingface/transformers/blob/eb3bd73ce35bfef56eeb722d697f2d39a06a8f8d/src/transformers/modeling_bart.py#L472), only very briefly and then it went back to `float32` - it seems that most layers like dropout or activation cast it right back into `float32`.

So `autocast` "works" too, albeit very differently from `apex`, where the data is `float16` in many more cases. If I go through the same `forward()` with apex, `x` is `float16` through and through.

Is this how `autocast` supposed to work? It appears so different from `apex`. Does it mean that all those operations not listed in the previous comment don't benefit from fp16 and have the same speed as fp32? Surely, at least it takes less space on the gpu.

**Comment by ptrblck** [comment]
> So I have verified that autocast "works" too, albeit very differently from apex, where both, the model weights and the intermediary data are float16 most of the time. 

`apex/amp` has different `opt_levels`, where the native amp implementation would be similar to `opt_level='O1'` in `apex`.

> Does it mean that all those operations not listed in the previous comment don't benefit from fp16 and have the same speed as fp32? Surely, at least it takes less space on gpu.

While one could reduce the memory footprint using FP16 in all operations, your training might blow up, as FP16 values can easily over/underflow. Automatic mixed-precision is designed to use FP32 where necessary, and FP16 where possible.
You can still use `model.half()` and use pure FP16.

Cross-posting from the huggingface repo:
https://github.com/huggingface/transformers/issues/8403#issuecomment-725356909

>After some more debugging it seems that the autocast cache is blowing up.
As a workaround you can add `torch.clear_autocast_cache()` in [BartForConditionalGeneration.forward](https://github.com/huggingface/transformers/blob/eb3bd73ce35bfef56eeb722d697f2d39a06a8f8d/src/transformers/modeling_bart.py#L1036), which might slow down your code but should at least work.
>Based on @stas00 's debugging it seems that PL is interacting with native AMP in a way that the cache is increasing.

**Comment by stas00** [comment]
Thank you for your answers, @ptrblck! 

I suppose the only way to really know which  amp/level would be most beneficial is to write a memory/speed benchmark.

**Comment by SeanNaren** [comment]
cross-posting... https://github.com/huggingface/transformers/issues/8403#issuecomment-725579899

> Don't think this is a trainer issue, I've been able to replicate this OOM crash putting autocast directly into the forward call, wrapping the code found here with torch.cuda.amp.autocast:
https://github.com/huggingface/transformers/blob/master/src/transformers/modeling_bart.py#L419
and turning off --fp16. Haven't been able to investigate further into why memory isn't being freed in this block!

**Comment by stas00** [comment]
Please let's use https://github.com/huggingface/transformers/issues/8403 for tracking the native amp issue, so that it's in one place.

**Comment by stas00** [comment]
I found another unrelated strange huge gpu memory requirement in PL. It happens during `save_checkpoint` - I observe the memory growing from 2GB (training was complete) to 5GB.

It's very difficult to debug PL due to dozens of indirections :( here is the stack trace (in v1.0.6)

```
dump_checkpoint, checkpoint_connector.py:337
save_checkpoint, checkpoint_connector.py:389
save_checkpoint, properties.py:207
_save_model, model_checkpoint.py:333
_update_best_and_save, model_checkpoint.py:583
_save_top_k_checkpoints, model_checkpoint.py:534
save_checkpoint, model_checkpoint.py:232
on_validation_end, model_checkpoint.py:186
on_validation_end, callback_hook.py:177
call_hook, trainer.py:833
on_evaluation_end, evaluation_loop.py:109
run_evaluation, trainer.py:620
run_training_epoch, training_loop.py:589
train, trainer.py:493
train_or_test, accelerator.py:74
train, gpu_accelerator.py:63
fit, trainer.py:444
generic_train, lightning_base.py:398
main, finetune.py:413
<module>, finetune.py:446
```
So the gpu ram is growing through `dump_checkpoint` about 1g, and then jumps by another 2GB at:
```
        # give the model a chance to dump a few things
        model.on_save_checkpoint(checkpoint)
```
Would this be something that can be done on CPU? This is saving a checkpoint - shouldn't need more gpu here in theory.

Ideally this function should need 0 extra gpu ram. Otherwise one will have a successful training but the program would OOM at the end of it during checkpoint saving which is not best.

**Comment by SeanNaren** [comment]
Update: This has been fixed in PyTorch: https://github.com/pytorch/pytorch/pull/48696 more info here: https://github.com/pytorch/pytorch/issues/48049#issuecomment-736701769

This will be fixed for Pytorch 1.8 and can be closed as this isn't to do with PL :)

**Comment by stas00** [comment]
Thank you, @SeanNaren!

**Comment by SeanNaren** [comment]
Thank you @stas00 for chasing this up, I think you got a pretty big bug resolved!
